### PR TITLE
Branch final ug dg

### PR DESCRIPTION
**Pull request by bingsen0806**

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1498,9 +1498,49 @@ testers are expected to do more *exploratory* testing.
...
+
+Another challenge is that the use case of LibTask also requires that all books with the same isbn must have the same name and authors. An initial consideration is to use the `Abstract Occurence` pattern in Week 10 topics, where a single book is represented by `Abstraction` and `Occurence` classes. However, this pattern binds book name and authors to isbn, and enforces a stronger than necessary requirement that books with same name must have the same isbn. This does not fit our use case because books with different edition can have the same name but different isbn. Hence, our team deliberately chose not to enforce such a constraint through design patterns, but rather introduced the following invariant and maintained the invariant at each step of command execution:
```
Comment (Halpfrog) [inline docs/DeveloperGuide.md:1509]: Well written explanation as to why we chose to take certain design choices.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1498,9 +1498,49 @@ testers are expected to do more *exploratory* testing.
...
+
+`book edit`: When a book is edited to another existing isbn, it may be an available book, which causes the new isbn to have an available copy. One way of maintaining invariant 1 is to remove all requesters for the new isbn and notify them, while another way is to disallow such edits. Our team chose the later method as this method is a more rational way of maintaining invariant 2 as well. When editing a book into another existing isbn, the set of requesters may be different for both old and new isbn, resulting in the new isbn now having a book copy with different requesters. For example, the user wants to change isbn of book A from 1 to 2, but different requesters are requesting for books with isbn 1 and 2. One method is to change the requesters of all books with isbn 2 to the requesters of book A, then invariant 2 is maintained. This does not make sense from the user point of view because the requesters of book A were originally requesting for isbn 1, not isbn 2. Hence, disallowing such an action is a better way to maintain data consistency.
```
Comment (Halpfrog) [inline docs/DeveloperGuide.md:1535]: Nice explanation for how we maintain data integrity

**Review by Halpfrog — COMMENTED** [review]
LGTM!

**Review by Halpfrog — APPROVED** [review]

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (yl-ang) [inline docs/DeveloperGuide.md:1509]: Minor typo `Occurence` should be `Occurrence`

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1498,9 +1498,49 @@ testers are expected to do more *exploratory* testing.
...
+
+Another challenge is that the use case of LibTask also requires that all books with the same isbn must have the same name and authors. An initial consideration is to use the `Abstract Occurence` pattern in Week 10 topics, where a single book is represented by `Abstraction` and `Occurence` classes. However, this pattern binds book name and authors to isbn, and enforces a stronger than necessary requirement that books with same name must have the same isbn. This does not fit our use case because books with different edition can have the same name but different isbn. Hence, our team deliberately chose not to enforce such a constraint through design patterns, but rather introduced the following invariant and maintained the invariant at each step of command execution:
+
+**Invariant 1**: If a book has same isbn as another book, then both books must have the same name, and same set of authors.
+
```
Comment (yl-ang) [inline docs/DeveloperGuide.md:1512]: Good invariant 💯

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1498,9 +1498,49 @@ testers are expected to do more *exploratory* testing.
...
+The original AB-3 code only contains one main type of object, which is `Person` object. Upon extension of `Model` component to include `Book` object that references `Patron` object, many data inconsistencies arise due to the immutable nature of `Book` and `Patron`. One example is that when a patron is edited, books borrowed and requested by that patron may not have their borrower and requester information updated, causing the name displayed on Book Card to be outdated.
+
+Instead of relying only on testing, LibTask uses a formal method approach to establish certain invariants to be held true after execution of every command, and use these invariants to prove the absence of data inconsistency bugs during the execution of the program. Some invariants were shown below to illustrate the careful consideration by the team to eliminate bugs:
+
+**Invariant 1**: A book can have requesters if and only if it does not have an available copy (i.e. another book with the same isbn).
+
+**Invariant 2**: Books with the same isbn must have the same set of requesters.
+
+**Invariant 3**: Borrower and requester information of a book must be the most updated patron information.
```
Comment (yl-ang) [inline docs/DeveloperGuide.md:1525]: Good inclusion of invariants to justify the addition of features 💯

**Review by yl-ang — APPROVED** [review]
LGTM! Good job with adding the effort section 💯

**Pull request merged**